components/ai_chat/core/browser/engine/engine_consumer_conversation_api.cc

```diff
@@ -308,6 +320,14 @@ void EngineConsumerConversationAPI::GenerateAssistantResponse(
        ++message_index) {
     const auto& message = conversation_history[message_index];
 
+    // Append associated content for the message (if any).
+    auto page_content_it = page_contents_messages.find(message->uuid.value());
+    if (page_content_it != page_contents_messages.end()) {
+      for (auto& event : page_content_it->second) {
+        conversation.emplace_back(std::move(event));
+      }
+    }
```

Comment:
Why can't we do the event creation here instead of step1? We are currently doing the search twice and I don't understand reason.

Reply:
We can't because we iterate over messages from oldest to newest and we need to prefer the newest page_content so when we run out of context we truncate the oldest page content first. 

This is similar to what we we do for tool_use_events.

https://github.com/brave/brave-core/pull/29627#discussion_r2167900307

Reply:
we need similar comments as in [oai.cc](https://github.com/brave/brave-core/pull/30487#discussion_r2270817904)